plugins/ua_securitypolicy_basic128rsa15.c

```diff
@@ -0,0 +1,1299 @@
+
+    retval = UA_ByteString_copy(certificate, &contextData->localCert);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    if(contextData->localCertThumbprint.length != 20) {
+        retval = UA_ByteString_allocBuffer(&contextData->localCertThumbprint,
+                                           20);
+
+        if(retval != UA_STATUSCODE_GOOD)
+            return retval;
+    }
+
+    retval = contextData->securityPolicy->asymmetricModule.
+        makeCertificateThumbprint(contextData->securityPolicy,
+                                  certificate,
+                                  &contextData->localCertThumbprint);
+
+    return retval;
+}
+
+static UA_StatusCode
+policyContext_setCertificateTrustList_sp_basic128rsa15(const UA_ByteString *const trustList,
+                                                       const size_t trustListCount,
+                                                       UA_SP_basic128rsa15_PolicyContextData *const contextData) {
+    if(trustList == NULL || contextData == NULL)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    for(size_t i = 0; i < trustListCount; ++i) {
+        int err = mbedtls_x509_crt_parse(&contextData->certificateTrustList,
+                                         trustList[i].data,
+                                         trustList[i].length);
+        // TODO: Do we want to keep parsing and only warn the user?
+        if(err) // TODO: more precise error handling?
+            return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+policyContext_setCertificateRevocationList_sp_basic128rsa15(const UA_ByteString *const revocationList,
+                                                            const size_t revocationListCount,
+                                                            UA_SP_basic128rsa15_PolicyContextData *const contextData) {
+    if(revocationList == NULL && contextData == NULL)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    for(size_t i = 0; i < revocationListCount; ++i) {
+        int err = mbedtls_x509_crl_parse_der(&contextData->certificateRevocationList,
+                                             revocationList[i].data,
+                                             revocationList[i].length);
+        // TODO: Do we want to keep parsing and only warn the user?
+        if(err) // TODO: more precise error handling?
+            return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+policyContext_newContext_sp_basic128rsa15(UA_SecurityPolicy *const securityPolicy,
+                                          const UA_ByteString localCertificate,
+                                          const UA_ByteString localPrivateKey,
+                                          const UA_ByteString *const certificateTrustList,
+                                          const size_t certificateTrustListCount,

```

Comment:
Other parts of the sdk will want access to the trust list as well. Notably session handling.
Pull out into a dedicated plugin.
How would the security policy interact with a key infrastructure plugin?

Reply:
What i had in mind was a plugin that can be queried for certificates and certificate validation.
The security policy can still choose to store a copy of the certificate in a parsed form or whatever, since it will only affect the implementation .c file. The plugin would be initialized once by the user with the certificates (or maybe their location, if we store stuff file-based) and then passed to the server and security policies.
Each security policy would then have a pointer to the plugin with which it can access the certificates and verify them via the plugin API.

Another thing to consider is where to store/load the certificates (and also the configuration, which is currently hard coded). This could maybe be handled by a plugin as well. (file system might not always be an option, right?)

Reply:
Although i would like to put such a plugin into a separate PR. Either before or after this. It shouldn't be much of a problem to change the current implementation to work with a plugin for certificates.